### PR TITLE
Using clusterrole instead of current registration roles

### PR DESCRIPTION
**Pull request by skeeey**
Signed-off-by: liuwei <liuweixa@redhat.com>

### CODE REVIEW

**Comment by skeeey** [comment]
refer to https://github.com/open-cluster-management/backlog/issues/9057

File: pkg/hub/clusterrole/controller.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	clusterInformer clusterv1informer.ManagedClusterInformer,
+	clusterRoleInformer rbacv1informers.ClusterRoleInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &clusterroleController{
+		kubeClient:    kubeClient,
+		clusterLister: clusterInformer.Lister(),
+		eventRecorder: recorder.WithComponentSuffix("managed-cluster-clusterrole-controller"),
+	}
+	return factory.New().
+		WithInformers(clusterInformer.Informer(), clusterRoleInformer.Informer()).
```
Comment (qiujian16) [inline pkg/hub/clusterrole/controller.go]: adding filter here will be better.
I think you also need clusterinformer?

When it is triggered when we create a new cluster?
When it is triggered when there are no clusters.

Comment (skeeey) [reply pkg/hub/clusterrole/controller.go]: OK, I will adding a filer here and I already added the `clusterinformer`

File: pkg/hub/clusterrole/controller.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+			func(name string) ([]byte, error) {
+				return assets.MustCreateAssetFromTemplate(name, bindata.MustAsset(filepath.Join(manifestDir, name)), nil).Data, nil
+			},
+			clusterRoleFiles...,
+		)
+	}
+
+	// Make sure the managedcluser cluserroles are existed if there are clusters
+	results := resourceapply.ApplyDirectly(
+		resourceapply.NewKubeClientHolder(c.kubeClient),
```
Comment (qiujian16) [inline pkg/hub/clusterrole/controller.go:91]: should we consider getting clusterrole/binding from lister at first?

Comment (skeeey) [reply pkg/hub/clusterrole/controller.go:91]: yeah, I think so, I will update this

File: pkg/hub/clusterrole/controller.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+			clusterRoleFiles...,
+		)
+	}
+
+	// Make sure the managedcluser cluserroles are existed if there are clusters
+	appliedClusterRoleFiles := []string{}
+	_, err = c.clusterRoleLister.Get(registrationClusterRole)
+	switch {
+	case errors.IsNotFound(err):
+		appliedClusterRoleFiles = append(appliedClusterRoleFiles, clusterRoleFiles[0])
```
Comment (qiujian16) [inline pkg/hub/clusterrole/controller.go]: do not use index here, you can have two consts for each file.

File: pkg/hub/clusterrole/controller.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	case err != nil:
+		return err
+	}
+
+	_, err = c.clusterRoleLister.Get(workClusterRole)
+	switch {
+	case errors.IsNotFound(err):
+		appliedClusterRoleFiles = append(appliedClusterRoleFiles, clusterRoleFiles[1])
+	case err != nil:
+		return err
```
Comment (qiujian16) [inline pkg/hub/clusterrole/controller.go]: should we also compare the rules in clusterroles?

File: pkg/hub/clusterrole/controller.go (hunk 2 above)
Comment (qiujian16) [reply pkg/hub/clusterrole/controller.go:91]: i will take it back...we did not able to compare the roles with lister. I think we should start with Get at first.

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**